experimental/formatted_string/src/formatted_string.rs

```diff
@@ -0,0 +1,219 @@
+    bytes: Vec<u8>,
+    // The first dimension corresponds to the bytes, the second are the L levels of annotations
+    annotations: Vec<[(LocationInPart, F); L]>,
+}
+
+impl<F: Copy, const L: usize> AsRef<str> for FormattedString<F, L> {
+    fn as_ref(&self) -> &str {
+        unsafe { &str::from_utf8_unchecked(&self.bytes) }
+    }
+}
+
+impl<F: Copy, const L: usize> FormattedStringLike<F, L> for FormattedString<F, L> {
+    fn annotation_at(&self, pos: usize) -> &[(LocationInPart, F); L] {
+        &self.annotations[pos]
+    }
+}
+
+impl<F: Copy, const L: usize> FormattedString<F, L> {
+    pub fn new() -> Self {
+        // A FormattedString with 0 annotations doesn't make sense.
+        assert!(L > 0);
+        Self {
+            bytes: Vec::with_capacity(40),
+            annotations: Vec::with_capacity(40),
+        }
+    }
+
+    pub fn append<S, const L1: usize>(&mut self, string: &S, field: F) -> &mut Self
+    where
+        S: FormattedStringLike<F, L1>,
+    {
+        assert_eq!(L - 1, L1);
+        // len() is always a char boundary
+        self.insert_internal(self.bytes.len(), string, field)
+    }
+
+    pub fn prepend<S, const L1: usize>(&mut self, string: &S, field: F) -> &mut Self
+    where
+        S: FormattedStringLike<F, L1>,
+    {
+        assert_eq!(L - 1, L1);
+        // 0 is always a char boundary
+        self.insert_internal(0, string, field)
+    }
+
+    pub fn insert<S, const L1: usize>(
+        &mut self,
+        pos: usize,
+        string: &S,
+        field: F,
+    ) -> Result<&mut Self, FormattedStringError>
+    where
+        S: FormattedStringLike<F, L1>,
+    {
+        assert_eq!(L - 1, L1);
+        if pos > self.bytes.len() {
+            return Err(FormattedStringError::IndexOutOfBounds(pos));
+        }
+        // bytes is valid UTF-8 precisely because we do this check before
+        // insertion (and string is valid UTF-8)
+        let current = unsafe { str::from_utf8_unchecked(&self.bytes) };
+        if !current.is_char_boundary(pos) {
+            Err(FormattedStringError::PositionNotCharBoundary(
+                pos,
+                current.to_owned(),
+            ))
+        } else {
+            Ok(self.insert_internal(pos, string, field))
+        }
+    }
+
+    // Precondition here is that pos is a char boundary and < buffer_len.
+    fn insert_internal<S, const L1: usize>(&mut self, pos: usize, string: &S, field: F) -> &mut Self
+    where
+        S: FormattedStringLike<F, L1>,
+    {
+        assert_eq!(L - 1, L1);
+        self.bytes.splice(pos..pos, string.as_ref().bytes());
+        self.annotations.splice(
+            pos..pos,
+            (0..string.as_ref().len()).map(|i| {
+                let top_level = (
+                    if i == 0 {
+                        LocationInPart::Begin
+                    } else {
+                        LocationInPart::Extend
+                    },
+                    field,
+                );
```

Comment:
Note: I think we likely will want to revise this logic a bit, but let's wait until we have more users like DTF and NF.  For example, when putting together a formatted number, several digits might be appended separately, but they should all coalesce into the same field.